nifi-nar-bundles/nifi-update-attribute-bundle/nifi-update-attribute-processor/src/main/java/org/apache/nifi/processors/attributes/UpdateAttribute.java

```diff
@@ -474,38 +496,49 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
+                // apply the actions for each rule and transfer the flowfile
+                for (final Map.Entry<FlowFile, List<Rule>> entry : matchedRules.entrySet()) {
+                    FlowFile match = entry.getKey();
+                    final List<Rule> rules = entry.getValue();
+                    boolean updateWorking = incomingFlowFile.equals(match);
+
+                    // execute each matching rule(s)
+                    match = executeActions(session, context, rules, defaultActions, match, stateInitialAttributes, stateWorkingAttributes);
+
+                    if (updateWorking) {
+                        incomingFlowFile = match;
+                    }
 
-                // execute each matching rule(s)
-                match = executeActions(session, context, rules, defaultActions, match, stateInitialAttributes, stateWorkingAttributes);
+                    if (debugEnabled) {
+                        logger.debug("Updated attributes for {}; transferring to '{}'", new Object[]{match, REL_SUCCESS.getName()});
+                    }
 
-                if (updateWorking) {
-                    incomingFlowFile = match;
+                    // add the match to the list to transfer
+                    flowFilesToTransfer.add(match);
                 }
+            } else {
+                // Either we're running without any rules or the FlowFile didn't match any
+                incomingFlowFile = executeActions(session, context, null, defaultActions, incomingFlowFile, stateInitialAttributes, stateWorkingAttributes);
 
                 if (debugEnabled) {
-                    logger.debug("Updated attributes for {}; transferring to '{}'", new Object[]{match, REL_SUCCESS.getName()});
+                    logger.debug("Updated attributes for {}; transferring to '{}'", new Object[]{incomingFlowFile, REL_SUCCESS.getName()});
                 }
 
-                // add the match to the list to transfer
-                flowFilesToTransfer.add(match);
+                // add the flowfile to the list to transfer
+                flowFilesToTransfer.add(incomingFlowFile);
             }
-        } else {
-            // Either we're running without any rules or the FlowFile didn't match any
-            incomingFlowFile = executeActions(session, context, null, defaultActions, incomingFlowFile, stateInitialAttributes, stateWorkingAttributes);
+        } catch (Exception e){
+            final String failureAction = context.getProperty(FAILURE_ACTION).getValue();
 
-            if (debugEnabled) {
-                logger.debug("Updated attributes for {}; transferring to '{}'", new Object[]{incomingFlowFile, REL_SUCCESS.getName()});
+            if(failureAction.equals(FAILURE_ACTION_ROLLBACK)) {
+                throw e;
             }
 
-            // add the flowfile to the list to transfer
-            flowFilesToTransfer.add(incomingFlowFile);
+            session.penalize(incomingFlowFile);
```

Comment:
I don't think flowfiles being sent to the failure relationship should be penalized. I think this line should move up to Line 537 inside the `if (failureAction.equals(FAILURE_ACTION_ROLLBACK)) {` block.

Reply:
I'll remove it from here. In a rollback scenario, penalization is handled automatically at the session level by throwing the exception (you'll note it is not something I moved/removed).